presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java

```diff
@@ -149,9 +159,65 @@ public FunctionListBuilder scalar(Class<?> clazz)
         catch (IllegalAccessException e) {
             throw Throwables.propagate(e);
         }
+        groupMatchingScalars();
+        return this;
+    }
+
+    private FunctionListBuilder groupMatchingScalars()
+    {
+        List<ParametricFunction> newFunctions = groupMatchingScalars(functions);
+        functions.clear();
+        functions.addAll(newFunctions);
         return this;
     }
 
+    private List<ParametricFunction> groupMatchingScalars(List<ParametricFunction> inputFunctions)
+    {
+        inputFunctions.forEach(f -> checkArgument(f instanceof FunctionInfo, "expected FunctionInfo but got %s", f));
+        Map<Signature, List<FunctionInfo>> groupedFunctions = inputFunctions.stream()
+                .map(f -> (FunctionInfo) f)
+                .collect(Collectors.groupingBy(ParametricFunction::getSignature));
```

Comment:
not great. 
Depends on the place when we add spaces. 
Do you think it is worthy to fight it?
The other problem you cound try to fight is that every function must use same variables for literal arguments (e.g. p and s). And theoretically we could allow one to use p,s and other a,b.